Issue 1005: Adding a checksums map to the ArtifactEntry class

ogryniuk opened the pull request:
Fixed #1005.

1. Declared a checksums map variable and its getters and setters to [ArtifactEntry](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java)
2. Included the variable in toString method

carlspring commented:
Hi!

Thank you for your contribution!

I don't see where this is actually used. There will also have to be a test for this. Would you mind looking into this? Thanks! :)

Kind regards,

Martin

fuss86 commented on strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java:
```diff
@@ -38,6 +40,8 @@
     @ManyToMany(targetEntity = ArtifactTagEntry.class)
     private Set<ArtifactTag> tagSet;
 
+    private Map<String, String> checkSumsMap;
```
`checkSumsMap` I think that `checksums` as a name would be enough

carlspring replied:
+1

ogryniuk replied:
Understood, I will implement this change as soon as I figure out another thing about this issue (how to test it). Then I will put the whole le in a pull request update

ogryniuk commented:
> Hi!
> 
> Thank you for your contribution!
> 
> I don't see where this is actually used. There will also have to be a test for this. Would you mind looking into this? Thanks! :)
> 
> Kind regards,
> 
> Martin



> Hi!
> 
> Thank you for your contribution!
> 
> I don't see where this is actually used. There will also have to be a test for this. Would you mind looking into this? Thanks! :)
> 
> Kind regards,
> 
> Martin

ogryniuk commented:
@carlspring, How exactly should I test my contribution? One method of the test class [ArtifactEntryServiceTest](https://github.com/strongbox/strongbox/blob/536a13a05073223c429e75d351fff0871d71a3ed/strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceTest.java#L57) refers to the class I work on, ArtifactEntry. So, how should I proceed? Declare a map with some String values and then test whether ArtifactEntry returns a map with the same values? Will it be enough and if yes, where should I put those lines of code, inside the existing method saveEntityShouldWork?

ogryniuk commented:
Hello @carlspring. I think how exactly should I test my contribution? 

One method of the test class [ArtifactEntryServiceTest](https://github.com/strongbox/strongbox/blob/536a13a05073223c429e75d351fff0871d71a3ed/strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceTest.java#L57) refers to the class I work on, ArtifactEntry. So, how should I proceed? 

Declare a map with some String values and then test whether ArtifactEntry returns a map with the same values? Will it be enough and if yes, where should I put those lines of code, inside the existing method saveEntityShouldWork?

Thanks for your feedback

carlspring commented:
@sbespalov / @fuss86 ,

Please, advise!

fuss86 commented:
@ogryniuk you forgot to implement the most important part of #1005. It says that:
> We should store this map in the ArtifactEntry @Entity to be able to search the database by the checksum algorithm and value. 

You just added a map, you don't store anything :( You haven't tested database search by stored checksum.

ogryniuk commented:
@fuss86 The thing is I did not forget, I thought declaring a setter for that map would be enough to give a possibility to store it whenever a certain user decides to call the method. It means that I unfortunately do not know how to store the map correctly and consequently can not resolve this issue. Sorry

sbespalov commented:
> It means that I unfortunately do not know how to store the map correctly

@ogryniuk the proper place to do it is `AbstractRepositoryProvider.commit()` method. There is `ArtifactEntry` instance available there, you only need to do the following:
```
LayoutOutputStream los = StreamUtils.findSource(LayoutOutputStream.class, ctx.getStream());
artifactEntry.getCheckSumsMap().clear();
artifactEntry.getCheckSumsMap().putAll(los.getDigestMap());
```

sbespalov commented on strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/domain/ArtifactEntry.java:
```diff
@@ -38,6 +40,8 @@
     @ManyToMany(targetEntity = ArtifactTagEntry.class)
     private Set<ArtifactTag> tagSet;
 
+    private Map<String, String> checkSumsMap;
```
@ogryniuk I also suggest to initialize the property with `new HashMap<>()` instance here, instead of having `Optional.ofNullable` logic in getter method.

ogryniuk replied:
@sbespalov Another collection, Set, is also initialised through Optional. Will not following your suggestion break the consistency of the class where one variable (Set) would be declared in the getter and another one (Map) - immediately after declaration?

sbespalov replied:
Right, I had not noticed that field actually initialized in getter as well. 
Good catch @ogryniuk, thanks :)

sbespalov commented:
> done in the project strongbox-db

@ogryniuk  can you please link this PR here.

sbespalov commented:
> Hello @carlspring. I think how exactly should I test my contribution?

@ogryniuk you should implement a test method within `ArtifactManagementServiceImplTest` which will store artifact and then check that `ArtifactEntry` has expected checksums stored.

ogryniuk commented:
> > done in the project strongbox-db
> 
> @ogryniuk can you please link this PR here.

@sbespalov . Here is the link to the PR in the database project [https://github.com/strongbox/strongbox-db/pull/8](url) . First there was a commit with an empty file inside but then I fixed it by adding a correct liquibase property.

Thanks a lot for the comments on how to store. I mean really, thanks a lot!)

ogryniuk commented:
While pushing a new commit I saw that a GitHub diff became huge and unreadable. I'am sorry about it and when writing code in IDE I did not format anything on my own and changed nothing but added a new method.

ogryniuk commented:
@sbespalov and all others concerned,

While pushing a new commit I saw that my GitHub diff became huge and unreadable. I'am sorry about it. When writing code in IDE I did no reformating of my own and changed nothing there except of adding a new method. Thus, I have no idea why the commit came out so ugly.

In essence, the commit is about adding a method testChecksumsStorage to the test class ArtifactManagementServiceImplTest

carlspring commented:
@ogryniuk ,

You seem to ave committed a `diff` file called `command`, which is causing the noise. Could you please remove this? Thanks!

ogryniuk commented:
> @ogryniuk , You seem to have committed a `diff` file called `command`, which is causing the noise

I also noticed 'command' file in my git repository, I deleted and untracked it before committing. But GitHub still included it in the commit (: 

In fact, I do not know where did it come from in the first place. Before committing, to be up to date with commits history I updated my forked repository and then pulled it to my machine. Maybe, I did some mistake in the process...

sbespalov commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactManagementServiceImplTest.java:
```diff
@@ -712,11 +723,63 @@ public void testLastVersionManagement(TestInfo testInfo)
+        MutableRepository repository = mavenRepositoryFactory.createRepository(repositoryId);
+        repository.setLayout(Maven2LayoutProvider.ALIAS);
+
+        createRepositoryWithArtifacts(STORAGE0,
```
in your case you should use `createRepository` instead of `createRepositoryWithArtifacts` because the artifact will be deployed within test.

sbespalov commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactManagementServiceImplTest.java:
```diff
@@ -712,11 +723,63 @@ public void testLastVersionManagement(TestInfo testInfo)
+        String gavtc = "org.carlspring.strongbox:strongbox-utils:8.4:jar";
+
+        File repositoryBasedir = getRepositoryBasedir(STORAGE0, repositoryId);
+        generateArtifact(repositoryBasedir.getAbsolutePath(), gavtc);
```
you should use this generated adtifact below instead of `generateArtifactInputStream`. 
you also have checksum files generated here, so you can read this checksums and use them as `expectedChecksum` below.

ogryniuk replied:
> you should use this generated adtifact below instead of `generateArtifactInputStream`. 
@sbespalov 
1. The method `mavenArtifactManagementService.store` requires as parameter a variable of type InputStream. If I remove `generateArtifactInputStream`, I will have no parameter for '`store`' method: `generateArtifact` returns a type MavenArtifact and am not aware of how to get from this type to the type InputStream. Or maybe, in this particular case I do not have to invoke `mavenArtifactManagementService.store` at all?
> you also have checksum files generated here
2. How can I access the checksum generated here?

sbespalov replied:
Yes, you need an InputStream, and you can get it from file, which the generateArtifact method produce inside. Try to go inside the method and look what happened there.

sbespalov commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactManagementServiceImplTest.java:
```diff
@@ -712,11 +723,63 @@ public void testLastVersionManagement(TestInfo testInfo)
+
+        File repositoryDir = getRepositoryBasedir(STORAGE0, repositoryId);
+
+        InputStream is = generateArtifactInputStream(repositoryDir.toPath().getParent().toAbsolutePath().toString(),
```
this should be dropped, see above

sbespalov commented on strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/services/ArtifactManagementServiceImplTest.java:
```diff
@@ -712,11 +723,63 @@ public void testLastVersionManagement(TestInfo testInfo)
+
+        ArtifactEntry artifactEntryForTest = artifactEntryService.findOneArtifact(STORAGE0, repositoryId, artifactPath);
+
+        Field field = ArtifactManagementService.class.getDeclaredField("checksumCacheManager");
```
this should be dropped, see above

carlspring commented on strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/io/AbstractRepositoryProvider.java:
```diff
@@ -215,7 +216,11 @@ public void commit(RepositoryStreamWriteContext ctx) throws IOException
-        
+
+        LayoutOutputStream los = StreamUtils.findSource(LayoutOutputStream.class, ctx.getStream());
+        artifactEntry.getChecksums().clear();
```
Why is this invocation of `clear()` necessary? It seems redundant.

ogryniuk replied:
@carlspring Frankly speaking I do not know. @sbespalov advised to add it.

ogryniuk commented:
Closed in favour of https://github.com/strongbox/strongbox/pull/1181
